Add new snapshots command in ctr for getting an image's snapshot layer mounts

katiewasnothere opened the pull request:
This PR adds a new command to ctr's `snapshot` commands for retrieving the snapshot mounts of an image from a given snapshotter on a given platform. Adding this utility to the `snapshot` commands makes it clear what the purpose is and what the caller's responsibilities are (iow cleaning up the snapshot). 

This new command is helpful for functional/unit tests in shims because it allows users to retrieve the layer mounts that would be used for creating a new container. Users can then manipulate or use these layer mounts as needed to test specific scenarios.

As an example, in https://github.com/microsoft/hcsshim we have a functional tests package for unit testing [here](https://github.com/microsoft/hcsshim/tree/master/test/functional). However, in order for us to unit test our shim, we must manually start a container or VM with an image's layer mounts. We're currently relying on an older docker command [as seen here](https://github.com/microsoft/hcsshim/blob/2010d9a3eeb0a93b64faac63c39eff76fb12082c/test/functional/utilities/layerfolders.go#L28) to retrieve those. This PR would replace the out of date docker command.


Signed-off-by: Kathryn Baldauf <kabaldau@microsoft.com>

katiewasnothere commented:
@kevpar @ambarve @dcantah @anmaxvl fyi

anmaxvl commented on cmd/ctr/commands/images/mounts.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+var mountsCommand = cli.Command{
+	Name:      "mounts",
```
maybe `list-mounts` and `lsm` for alias? not sure what the convention is.

ambarve commented:
I think having two commands with such similar names but different purpose can be confusing and also doesn't follow the convention. For example with ctr snapshotter `mount` and `mounts` are just aliases for the same command. I think the better approach is to add a new `--list-only` flag to the image mount command. What do you think?

cpuguy83 commented:
Agree here, `--list-only` or `--dry-run` on the mount command. Or a combination like `--show --dry-run`... or always show the mounts and add `--dry-run`.

dcantah commented on cmd/ctr/commands/images/mounts.go:
```diff
@@ -0,0 +1,80 @@
+		if snapshotter == "" {
+			snapshotter = containerd.DefaultSnapshotter
+		}
+
```
Don't need this, you can just pass `context.String("snapshotter")` to `client.SnapshotService(snapshotter)` as it will return the default snapshotter if it can't find the one requested. Don't know of any snapshotter that's registered for empty string (I hope there isn't one 😄)

ambarve commented on cmd/ctr/commands/images/mount.go:
```diff
@@ -130,14 +137,21 @@ When you are done, use the unmount command.
+			commands.PrintAsJSON(mounts)
+		}
+
+		if !context.Bool("dry-run") {
```
Now if this command is run with a `dry-run` option we will still create a snapshot for it and that snapshot will continue to exist. I think we probably should remove the snapshot before exiting if `dry-run` is enabled.

katiewasnothere replied:
That's tricky because for our use case we don't actually want to mount or remove the snapshot since we intend to use it in our functional tests. Perhaps yet another flag could be added? :p

cpuguy83 replied:
Having a dry-run that has left-over side-effects doesn't seem right, though 😟

katiewasnothere commented:
@ambarve @cpuguy83 I reworked this PR to instead make a new command in the list of `snapshots` commands. I've updated the title and the description. @ambarve and I were discussing the previous design and we decided that what the goal of my changes were just weren't very compatible with that route. Particularly around the "dry-run" and side effects. This new design makes it more clear what's going on and what work the caller needs to do to clean up or not. It also avoids any complications with the `Mount.All` command on different platforms (ex: getting the snapshots for a linux image on windows).

ambarve commented on cmd/ctr/commands/snapshots/snapshots.go:
```diff
@@ -271,6 +281,115 @@ var removeCommand = cli.Command{
+		},
+		cli.StringFlag{
+			Name:  "key",
+			Usage: "Optional key to use for snapshot. Default is image name.",
```
I think it might be okay to make this a required parameter just like the `prepare` command. Unless there is a specific use case where we don't care about the key with which the snapshot is created.

katiewasnothere replied:
Yeah that might be a good idea so that we can make sure people know what key to remove later

katiewasnothere replied:
Done :)

ambarve commented on cmd/ctr/commands/snapshots/snapshots.go:
```diff
@@ -271,6 +281,115 @@ var removeCommand = cli.Command{
+		}
+
+		s := client.SnapshotService(snapshotter)
+		mounts, err := s.Prepare(ctx, key, chainID)
```
I think we should provide the option of creating a read-only (View) snapshot too. We can add a `--read-only` flag for the command.

katiewasnothere replied:
Done :)

mxpv commented on cmd/ctr/commands/snapshots/snapshots.go:
```diff
@@ -271,6 +281,121 @@ var removeCommand = cli.Command{
+	ArgsUsage: "[flags] <key> <image name>",
+	Description: `Creates a new snapshot from a specified image.
+
+When you're done, use the remove command.
```
We already have `prepareCommand` and `viewCommand` for creating snapshots. Would it make sense to extend those instead? This one seems like does the same, but offers more CLI flags.

katiewasnothere commented:
> We already have `prepareCommand` and `viewCommand` for creating snapshots. Would it make sense to extend those instead? This one seems like does the same, but offers more CLI flags. 

I thought about that too, but I'm not sure how much we're okay with changing the expectations for those commands. The goal of this new command is really to take an image name, like "docker.io/library/alpine:latest", and get a new snapshot created with the mounts printed to stdout. `Prepare` and `View` both start out with image key and optionally the chain ID of the parent. I think there could be some refactoring done, but I'm not sure if it would make sense to try to merge them together. Thoughts?

katiewasnothere commented:
@mxpv that sounds reasonable to me. In that case, I'll close this PR and open a new one with the suggested change instead.
